rusty_pizza_server/src/order_model/order.rs

```diff
@@ -227,4 +319,339 @@ mod tests {
+        let mut order = Order::new(manager);
+
+        for attributes in meals_attributes.into_iter() {
+            let user = Rc::new(User::new(attributes.1));
+            order.add_user(user.clone());
+            order.set_tip_for_user(user.clone(), attributes.2).unwrap();
+        }
+        //When
+        let calculated_tip = order.calculate_total_tip();
+
+        //Then
+        assert_eq!(total_tip, calculated_tip);
+    }
+
+    #[rstest(meals_attributes, expected_change,
+        case(
+            vec![
+                MealsAttributes(
+                    vec![Money::new(2, 25), Money::new(5, 50), Money::new(7, 37)],
+                    String::from("Gabriel"),
+                    Money::new(17, 00),
+                ),
+                MealsAttributes(
+                    vec![Money::new(3, 50), Money::new(4, 42)],
+                    String::from("Michael"),
+                    Money::new(8, 50),
+                ),
+                MealsAttributes(
+                    vec![Money::new(6, 83)],
+                    String::from("Uriel"),
+                    Money::new(6, 83),
+                ),
+            ],
+            Money::new(2, 46),
+        ),
+        case(
+            vec![
+                MealsAttributes(
+                    vec![Money::new(2, 25), Money::new(4, 42)],
+                    String::from("Adam"),
+                    Money::new(8, 25),
+                ),
+                MealsAttributes(
+                    vec![Money::new(5, 50)],
+                    String::from("Eva"),
+                    Money::new(5, 50),
+                )
+            ],
+            Money::new(1, 58),
+        ),
+    )]
+    fn all_paid_enough_change_is_calculated_correctly(
+        meals_attributes: Vec<MealsAttributes>,
+        expected_change: Money,
+    ) {
+        //Given
+        let manager = Rc::new(User::new(String::from("Gott")));
+        let mut order = Order::new(manager);
+
+        for attributes in meals_attributes.into_iter() {
+            let user = Rc::new(User::new(attributes.1));
+            order.add_user(user.clone());
+            for price in attributes.0.iter() {
+                order
+                    .add_meal_for_user(
+                        user.clone(),
+                        String::from("XX"),
+                        String::from("something"),
+                        price.clone(),
+                    )
+                    .unwrap();
+            }
+            order.set_paid_for_user(user.clone(), attributes.2).unwrap();
+        }
+        //When
+        let calculated_change = order.calculate_total_change().unwrap();
+        //Then
+        assert_eq!(expected_change, calculated_change);
+    }
+
+    fn build_paid_less_hash_set(names: Vec<String>) -> HashSet<Rc<User>> {
+        let mut paid_less: HashSet<Rc<User>> = HashSet::new();
+        for name in names {
+            paid_less.insert(Rc::new(User::new(name)));
+        }
+        return paid_less;
```

Comment:
You could use `collect` here which would strip the `HashSet` construction:
```rust
names.iter().map(|name| Rc::new(User::new(name))).collect()
```

Reply:
I tried this form directly in the test attribute definiton, but cargo marked it as an error. So I extracted it to a method. Within in the method I see no benefit in the short form.

Reply:
In my opinion this makes the code clearer because it strips unnecessary technical details. Still, you are right that it is an encapsulated detail.

Reply:
Changed